erts/emulator/nifs/common/prim_file_nif.c

```diff
@@ -1399,3 +1404,75 @@ static ERL_NIF_TERM altname_nif(ErlNifEnv *env, int argc, const ERL_NIF_TERM arg
 
     return enif_make_tuple2(env, am_ok, result);
 }
+
+static ERL_NIF_TERM copy_file_available_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
+    off64_t ret;
+
+    ret = syscall(SYS_copy_file_range, -1, NULL, -1, NULL, 0, 0);
+
+    if (ret == -1 && errno == ENOSYS) {
+        return enif_make_atom(env, "false");
+    } else {
+        return enif_make_atom(env, "true");
+    }
+}
+
+static ERL_NIF_TERM copy_file_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
+    posix_errno_t posix_errno, ignored;
+    efile_data_t *d_in, *d_out;
+    off64_t ret = 0, len;
+    efile_fileinfo_t info = {0};
+    efile_path_t file_path_in, file_path_out;
+
+    ASSERT(argc == 2);
+
+    if ((posix_errno = efile_marshal_path(env, argv[0], &file_path_in))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+
+    if ((posix_errno = efile_marshal_path(env, argv[1], &file_path_out))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
+
+    /* We need to open Source file in order to get file size*/
+    if ((posix_errno = efile_open(&file_path_in, EFILE_MODE_READ, efile_resource_type, &d_in))) {
+        return posix_error_to_tuple(env, posix_errno);
+    }
```

Comment:
It would be better if this was entirely moved to the platform-specific implementations, e.g. `efile_copy_file` taking paths instead of file handles as arguments. You would then be able to use plain `open(2)` instead of the more clunky `efile_open`, or `CopyFileW` on Windows.

As a fallback for when `copy_file_range(2)` is unavailable you could have a simple `read(2)+write(2)` loop.